lightning/src/ln/interactivetxs.rs

```diff
@@ -0,0 +1,773 @@
+
+#[derive(Debug)]
+enum StateMachine {
+	Indeterminate,
+	LocalChange(LocalChange),
+	RemoteChange(RemoteChange),
+	LocalTxComplete(LocalTxComplete),
+	RemoteTxComplete(RemoteTxComplete),
+	NegotiationComplete(NegotiationComplete),
+	NegotiationAborted(NegotiationAborted),
+}
+
+impl Default for StateMachine {
+	fn default() -> Self {
+		Self::Indeterminate
+	}
+}
+
+macro_rules! define_state_machine_transitions {
+	($transition: ident, $msg: ty, [$(FROM $from_state: ident, TO $to_state: ident),+]) => {
+		fn $transition(self, msg: $msg) -> StateMachine {
+			match self {
+				$(
+					Self::$from_state(s) => match s.transition(msg) {
+						Ok(new_state) => StateMachine::$to_state(new_state),
+						Err(abort_reason) => StateMachine::NegotiationAborted(NegotiationAborted(abort_reason)),
+					}
+				 )*
+				_ => StateMachine::NegotiationAborted(NegotiationAborted(AbortReason::UnexpectedCounterpartyMessage)),
+			}
+		}
+	};
+	(LOCAL_OR_REMOTE_CHANGE, $to_local_transition: ident, $to_remote_transition: ident, $msg: ty) => {
+		define_state_machine_transitions!($to_local_transition, $msg, [
+			FROM RemoteChange, TO LocalChange,
+			FROM RemoteTxComplete, TO LocalChange
+		]);
+		define_state_machine_transitions!($to_remote_transition, $msg, [
+			FROM LocalChange, TO RemoteChange,
+			FROM LocalTxComplete, TO RemoteChange
+		]);
+	};
+}
+
+impl StateMachine {
+	fn new(feerate_sat_per_kw: u32, is_initiator: bool, tx_locktime: PackedLockTime) -> Self {
+		let context = NegotiationContext {
+			tx_locktime,
+			holder_is_initiator: is_initiator,
+			received_tx_add_input_count: 0,
+			received_tx_add_output_count: 0,
+			inputs: HashMap::new(),
+			prevtx_outpoints: HashSet::new(),
+			outputs: HashMap::new(),
+			feerate_sat_per_kw,
+		};
+		if is_initiator {
+			Self::RemoteChange(RemoteChange(context))
+		} else {
+			Self::LocalChange(LocalChange(context))
+		}
+	}
+
+	define_state_machine_transitions!(
+		LOCAL_OR_REMOTE_CHANGE, local_tx_add_input, remote_tx_add_input, &msgs::TxAddInput
+	);
+	define_state_machine_transitions!(
+		LOCAL_OR_REMOTE_CHANGE, local_tx_add_output, remote_tx_add_output, &msgs::TxAddOutput
+	);
+	define_state_machine_transitions!(
+		LOCAL_OR_REMOTE_CHANGE, local_tx_remove_input, remote_tx_remove_input, &msgs::TxRemoveInput
+	);
+	define_state_machine_transitions!(
+		LOCAL_OR_REMOTE_CHANGE, local_tx_remove_output, remote_tx_remove_output, &msgs::TxRemoveOutput
+	);
+	define_state_machine_transitions!(local_tx_complete, &msgs::TxComplete, [
+		FROM RemoteChange, TO LocalTxComplete,
+		FROM RemoteTxComplete, TO NegotiationComplete
+	]);
+	define_state_machine_transitions!(remote_tx_complete, &msgs::TxComplete, [
+		FROM LocalChange, TO RemoteTxComplete,
+		FROM LocalTxComplete, TO NegotiationComplete
+	]);
+}
+
+pub struct InteractiveTxConstructor<ES: Deref>
+	where
+		ES::Target: EntropySource,
+{
+	state_machine: StateMachine,
+	channel_id: [u8; 32],
+	is_initiator: bool,
+	entropy_source: ES,
```

Comment:
For me it's not a big deal if it bubbles up, but I like the idea of pregenerating them in "new()" and then only needing a reference to new, like channel constructors already use.

Reply:
Pre-generation, or providing inputs/outputs together with a serial ID sounds a simpler solution, but cannot cope well with the (unlikely) case of conflict with an entry provided by the counterparty (e.g. we pre-generate 1 and 2, send 1, receive 2, send 2). Also, in that case the constructor should check for uniqueness of the provided serial IDs.

Would it be possible to use a deterministic scheme for serial IDs? In that case there is no need to provide IDs upfront nor for entropy. E.g. few bytes of hash of channel_id + funding_tx_id + index + input/output hash, with extra extra increments to avoid eventual conflicts with already used IDs from both sides. Or event just smallest number that does not conflict with IDs used so far, would that work?

Reply:
I believe that using random serial IDs makes future multi-party protocols less brittle to deal with.